fem/gridfunc.cpp

```diff
@@ -2476,37 +2517,72 @@ void GridFunction::ProjectCoefficient(VectorCoefficient &vcoeff)
          }
          SetSubVector(vdofs, vals);
       }
-
    }
-
    else
    {
-      // Define and assemble linear form
-      LinearForm b(fes);
-      b.AddDomainIntegrator(new VectorFEDomainLFIntegrator(vcoeff));
-      b.Assemble();
-
-      // Define and assemble bilinear form
-      BilinearForm a(fes);
-      a.AddDomainIntegrator(new VectorFEMassIntegrator());
-      a.Assemble();
-
-      // Set solver and preconditioner
-      SparseMatrix A(a.SpMat());
-      GSSmoother  prec(A);
-      CGSolver cg;
-      cg.SetOperator(A);
-      cg.SetPreconditioner(prec);
-      cg.SetRelTol(1e-12);
-      cg.SetMaxIter(1000);
-      cg.SetPrintLevel(0);
-
-      // Solve and get solution
-      *this = 0.0;
-      cg.Mult(b,*this);
+      constexpr real_t signal = std::numeric_limits<real_t>::min();
+      for (int i = 0; i < fes->GetNE(); i++)
+      {
+         doftrans = fes->GetElementVDofs(i, vdofs);
+         vals.SetSize(vdofs.Size());
+         vals = signal;
+         fes->GetFE(i)->Project(vcoeff, *fes->GetElementTransformation(i), vals);
+         if (doftrans)
+         {
+            doftrans->TransformPrimal(vals);
+         }
+         // Remove undefined dofs
+         // The knot location (either Botella, Demko or Greville point)
+         // where the NURBS dof are evaluated might fall outside of the
+         // domain of the element. In that case the value is not set, and
+         // the value remains the signal value.
+         int s = 0;
+         for (int ii = 0; ii < vals.Size(); ii++)
+         {
+            if (vals[ii] != signal)
+            {
+               vdofs[s] = vdofs[ii];
+               vals(s) = vals(ii);
+               s++;
+            }
+         }
+         vdofs.SetSize(s);
+         vals.SetSize(s);
+
+         // Add reduced dofs to global vector
+         SetSubVector(vdofs, vals);
+      }
    }
 }
 
+void GridFunction::ProjectCoefficientGlobalL2(VectorCoefficient &vcoeff,
+                                              real_t rtol, int iter)
+{
+   // Define and assemble linear form
+   LinearForm b(fes);
+   b.AddDomainIntegrator(new VectorFEDomainLFIntegrator(vcoeff));
+   b.Assemble();
+
+   // Define and assemble bilinear form
+   BilinearForm a(fes);
+   a.AddDomainIntegrator(new VectorFEMassIntegrator());
+   a.Assemble();
+
+   // Set solver and preconditioner
+   SparseMatrix A(a.SpMat());
+   GSSmoother  prec(A);
+   CGSolver cg;
+   cg.SetOperator(A);
+   cg.SetPreconditioner(prec);
+   cg.SetRelTol(1e-12);
```

Comment:
`rtol, iter`

Reply:
I should check my sown stuff more/better...